scene/main/node.cpp

```diff
@@ -2618,6 +2580,62 @@ Node *Node::_duplicate(int p_flags, HashMap<const Node *, Node *> *r_duplimap) c
 		}
 	}
 
+	for (List<const Node *>::Element *N = node_tree.front(); N; N = N->next()) {
+		Node *current_node = node->get_node(get_path_to(N->get()));
+		ERR_CONTINUE(!current_node);
+
+		if (p_flags & DUPLICATE_SCRIPTS) {
+			bool is_valid = false;
+			Variant scr = N->get()->get(script_property_name, &is_valid);
+			if (is_valid) {
+				current_node->set(script_property_name, scr);
+			}
+		}
+
+		List<PropertyInfo> plist;
+		N->get()->get_property_list(&plist);
+
+		for (const PropertyInfo &E : plist) {
+			if (!(E.usage & PROPERTY_USAGE_STORAGE)) {
+				continue;
+			}
+			String name = E.name;
+			if (name == script_property_name) {
+				continue;
+			}
+
+			Variant value = N->get()->get(name).duplicate(true);
+
+			if (E.usage & PROPERTY_USAGE_ALWAYS_DUPLICATE) {
+				Resource *res = Object::cast_to<Resource>(value);
+				if (res) { // Duplicate only if it's a resource
+					current_node->set(name, res->duplicate());
+				}
+
+			} else {
+				// If property points to a node which is owned by a node we are duplicating, update its path.
+				if (value.get_type() == Variant::OBJECT) {
+					Node *property_node = Object::cast_to<Node>(value);
+					if (property_node && is_ancestor_of(property_node)) {
+						value = current_node->get_node_or_null(get_path_to(property_node));
+					}
+				} else if (value.get_type() == Variant::ARRAY) {
+					Array arr = value;
+					if (arr.get_typed_builtin() == Variant::OBJECT) {
```

Comment:
This won't handle non-typed Arrays, which can hold Objects too, but I guess it's not that important (and the current code is more performant).

Reply:
Do you want me to take a look at that before merging this pr?

Reply:
Unless you are talking about a case when you duplicate a node in gdscript and you use non-typed array to store nodes :thinking:

Reply:
If it can't be assigned then ok.

Reply:
I tested quickly assigning from code with non typed array and non typed variable, duplication seems to work correctly:

![image](https://github.com/godotengine/godot/assets/33091666/1c1bd6cf-4d53-4325-8495-943486ab7b17)


```
extends Node3D

@export var health: float = 100
@export var cube: Node3D = null
@export var cubes: Array[Node3D] = []
var non_typed_array: Array = []
var non_typed_node = null

# Called when the node enters the scene tree for the first time.
func _ready() -> void:
	health = PlayerStats.player_health
	non_typed_node = get_node("MeshInstance3D")
	non_typed_array.push_back(get_node("MeshInstance3D"))
	if name == "Cube":
		var dup_node: Node3D = duplicate()
		get_tree().get_current_scene().add_child.call_deferred(dup_node)
		dup_node.global_position.x += 3
	print(non_typed_array[0])

# Called every frame. 'delta' is the elapsed time since the previous frame.
func _physics_process(delta: float) -> void:
	if Input.is_action_just_pressed("hurt_self"):
		health -= 10
		PlayerStats.player_health = health
		print(health)
	if Input.is_action_just_pressed('switch_level'):
		get_tree().change_scene_to_file("res://scene_2.tscn")
	if non_typed_array.size() > 0:
		non_typed_array[0].rotation.y += 90 * delta
```